### PR TITLE
Generated voucher not displaying in the cart summary for a customer

### LINKED ISSUE #19442: Generated voucher not displaying in the cart summary for a customer
#### Describe the bug

A clear and concise description of what went wrong.

#### Expected behavior

Vouchers that were generated by the module are not displayed in the cart summary. 

image

In voucher settings, setting "Highlight" is enabled

image

#### Steps to Reproduce

- Generate a new voucher to a specific customer with a code
- Add some product to the cart
- Check voucher in the cart summare

Voucher is not displayed in the cart summary.

We sale on addons store a [module that use this feature](https://addons.prestashop.com/fr/fidelisation-parrainage/7301-programme-de-fidelite-avance-loyalty-points.html). We need to show a coupon generated for the user. If not he will forget to use it if he want and he need get the code on his account page... To complicated. 

For us it's a regression. We work on a PR.

#### Additional information

* PrestaShop version: 1.7.7.x
* PHP version: N/A

### PR DESCRIPTION
**Pull request by florian-202**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Voucher not displayed if it is linked to a customer, see also #19442 
| Type?         | bug fix 
| Category?     | FO 
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #19442 .
| How to test?  | - Open a PS1.7.7 <br> - create a new voucher limited to 1 person with code code <br>- Add some product to the cart (you should be logged into account of the person selected for voucher) <br>- Check voucher in the cart summary <br> <br> Please check also that https://github.com/PrestaShop/PrestaShop/issues/13002#issuecomment-476516323 is still fixed as this PR modifies it

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/19443)

### CODE REVIEW

**Comment by matks** [comment]
Thanks !

It's weird 😅 I added this line from https://github.com/PrestaShop/PrestaShop/commit/d37c9b7f5944ae3b980b4375f489f9f26210038a to fix bug https://github.com/PrestaShop/PrestaShop/issues/13002

**Comment by matks** [comment]
I need to check, when following your scenario, what Customer ID is used 😮

**Review by matks — CHANGES_REQUESTED** [review]
Warning: you modified the file rights from 644 to 755 😉 

We need 644 for files and 755 for folders

**Comment by clotairer** [comment]
> I need to check, when following your scenario, what Customer ID is used 

Create a user in front and a new voucher in back associated to this user.

> We need 644 for files and 755 for folders

ping @florian-202

**Review by Progi1984 — APPROVED** [review]

**Review by matks — APPROVED** [review]
File rights fixed, thanks !

**Comment by Progi1984** [comment]
@florian-202 Could you rebase your PR to the branch `1.7.7.x` and target it in Github ?

**Comment by matks** [comment]
> @florian-202 Could you rebase your PR to the branch `1.7.7.x` and target it in Github ?

@florian-202 this might be a complex operation, please tell us if it's complicated and we'll do it 😉

File: src/Adapter/Presenter/Cart/CartPresenter.php [hunk 1]
```diff
@@ -448,7 +448,8 @@ function ($voucher) {
         $discounts = array_filter($discounts, function ($discount) use ($cartRulesIds, $cart) {
             $voucherCustomerId = (int) $discount['id_customer'];
             $voucherIsRestrictedToASingleCustomer = ($voucherCustomerId !== 0);
-            if ($voucherIsRestrictedToASingleCustomer && $cart->id_customer !== $voucherCustomerId) {
+            $voucherIsEmptyCode = empty($discount['code']);
+            if ($voucherIsRestrictedToASingleCustomer && $cart->id_customer !== $voucherCustomerId && $voucherIsEmptyCode) {
```
Comment (PierreRambaud) [inline src/Adapter/Presenter/Cart/CartPresenter.php:452]:
```suggestion
            if (($voucherIsRestrictedToASingleCustomer && $cart->id_customer !== $voucherCustomerId) || $voucherIsEmptyCode) {
```

Should it be something like this instead? :thinking: 
It looks weird to authorized `$voucherIsRestrictedToASingleCustomer = true` && `$cart->id_customer !== $voucherCustomerId` if the code is not empty :thinking:

Comment (clotairer) [reply src/Adapter/Presenter/Cart/CartPresenter.php:452]: In fact, 
- if $voucherIsRestrictedToASingleCustomer = true && $cart->id_customer !== $voucherCustomerId && code is empty, > it's a credit note. The voutcher is automaticly apply > You don't need to display the voucher if hightlight is set on it.
- if $voucherIsRestrictedToASingleCustomer = true && $cart->id_customer !== $voucherCustomerId && but code is NOT empty, >  it's a something like a fidelity voucher, a discount. Usage is not mandatory. > You need to display the voucher if hightlight is set on it.

Comment (PierreRambaud) [reply src/Adapter/Presenter/Cart/CartPresenter.php:452]: Thanks for explanation :wink:

Comment (clotairer) [reply src/Adapter/Presenter/Cart/CartPresenter.php:452]: Note that if
 $voucherIsRestrictedToASingleCustomer = false  && code is empty > I don't know the use case... Probably we need to display the voucher

**Comment by matks** [comment]
@clotaire202 @florian-202 hi ! I tried to backport this branch to 1.7.7.x but I could not update this PR and branch because you did not give maintainers the rights to edit it 😄 

So I used a 2nd solution: I backported your commit to another branch and created another PR => https://github.com/PrestaShop/PrestaShop/pull/19456

So the PR author is me but the code is still yours, as you can see in https://github.com/PrestaShop/PrestaShop/pull/19456/commits => you will be credited as the author

I also squashed your 5 commits into a single one as it was easier to cherry-pick this way

**Comment by Robin-Fischer-PS** [comment]
@matks Should I test it on develop, or wait for your PR 1.7.7.x to be reviewed ?

**Comment by matks** [comment]
Since https://github.com/PrestaShop/PrestaShop/pull/19456 has been merged in 1.7.7.x ; next merge from 1.7.7.x into develop will update develop branch with the work of @florian-202 so we dont need this PR anymore

Thanks ! @florian-202 !

**Pull request closed without merging**